dali/pipeline/pipeline_params.h

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2017-2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_PIPELINE_PARAMS_H_
+#define DALI_PIPELINE_PIPELINE_PARAMS_H_
+
+#include <optional>
+#include "dali/pipeline/executor/queue_metadata.h"
+#include "dali/pipeline/executor/executor_type.h"
+
+namespace dali {
+
+/** Pipeline parameters used on construction of the Pipeline object */
+struct PipelineParams {
+  std::optional<int> max_batch_size;
+  std::optional<int> num_threads;
+  std::optional<int> device_id;
+  std::optional<int64_t> seed;
+  std::optional<ExecutorType> executor_type;
+  std::optional<ExecutorFlags> executor_flags;
+  std::optional<QueueSizes> prefetch_queue_depths;
+  std::optional<bool> enable_checkpointing;
+  std::optional<bool> enable_memory_stats;
+  std::optional<size_t> bytes_per_sample_hint;
```

Comment:
Wouldn't this be a slight abuse of `optional`? My point is - I believe most (if not all) of these parameters are not optional, since their values are actually required for the Executor to operate. Instead `optional` - should they have good default values?

Reply:
The goal of the whole exercise was to be able to NOT provide a value, so we can keep whatever was serialized.
With defaults we don't have this possibility.
C API 2.0 uses a similar approach, just using bit fields `xxx_present` instead of `optional`.